scilpy/denoise/asym_filtering.py

```diff
@@ -1,387 +1,480 @@
+    Build the angle filter, weighted on angle between current direction u
+    and neighbour direction v.
 
     Parameters
     ----------
-    in_sh: ndarray (x, y, z, ncoeffs)
-        Input SH volume.
-    sh_order: int, optional
-        Maximum SH order of input volume.
-    sh_basis: str, optional
-        Name of SH basis used.
-    in_full_basis: bool, optional
-        True if input is expressed in full SH basis.
-    is_legacy : bool, optional
-        Whether or not the SH basis is in its legacy form.
-    sphere_str: str, optional
-        Name of the DIPY sphere to use for sh to sf projection.
-    sigma_spatial: float, optional
-        Standard deviation for spatial filter.
-    sigma_angular: float, optional
-        Standard deviation for angular filter.
-    sigma_range: float, optional
-        Standard deviation for range filter.
+    sigma_angle: float
+        Standard deviation of filter. Values at distances greater than
+        sigma_angle are clipped to 0 to reduce computation time.
+    sphere: DIPY sphere
+        Sphere used for sampling the SF.
 
     Returns
     -------
-    out_sh: ndarray (x, y, z, ncoeffs)
-        Output SH coefficient array in full SH basis.
+    weights: ndarray
+        Angle filter of shape (N_dirs, N_dirs).
     """
-    # Load the sphere used for projection of SH
-    sphere = get_sphere(sphere_str)
-
-    # Normalized filter for each sf direction
-    s_weights = _get_spatial_weights(sigma_spatial)
-    a_weights = _get_angular_weights(s_weights.shape, sphere, sigma_angular)
-
-    weights = s_weights[..., None] * a_weights
-    weights /= np.sum(weights, axis=(0, 1, 2))
-
-    nb_sf = len(sphere.vertices)
-    B = sh_to_sf_matrix(sphere, sh_order=sh_order, basis_type=sh_basis,
-                        return_inv=False, full_basis=in_full_basis,
-                        legacy=is_legacy)
-
-    mean_sf = np.zeros(in_sh.shape[:-1] + (nb_sf,))
-
-    # Apply filter to each sphere vertice
-    for sph_id in range(nb_sf):
-        w_filter = weights[..., sph_id]
-
-        # Generate 1-channel images for directions u and -u
-        current_sf = np.dot(in_sh, B[:, sph_id])
-        mean_sf[..., sph_id] = _correlate_spatial(current_sf, w_filter,
-                                                  sigma_range)
-
-    # Convert back to SH coefficients
-    _, B_inv = sh_to_sf_matrix(sphere, sh_order=sh_order, basis_type=sh_basis,
-                               full_basis=True, legacy=is_legacy)
-    out_sh = np.array([np.dot(i, B_inv) for i in mean_sf], dtype=in_sh.dtype)
-    # By default, return only asymmetric SH
-    return out_sh
+    directions = sphere.vertices
+    if sigma_angle is not None:
+        dot = directions.dot(directions.T)
+        x = np.arccos(np.clip(dot, -1.0, 1.0))
+        weights = _evaluate_gaussian_distribution(x, sigma_angle)
+        mask = x > (3.0*sigma_angle)
+        weights[mask] = 0.0
+        weights /= np.sum(weights, axis=-1)
```

Comment:
Is there any possibility that you divide by 0 here?

Reply:
yes if an invalid sigma_angle < 0 is given. I'll add checks at the beginning of the calling method.